contracts/instance/InstanceAccessManager.sol

```diff
@@ -1,292 +1,526 @@
+        (bool inRole, uint32 executionDelay) = _accessManager.hasRole(admin.toInt(), _msgSender());
+        assert(executionDelay == 0); // to be sure no delayed execution functionality is used
+        if (!inRole) {
+            revert IAccess.ErrorIAccessCallerIsNotRoleAdmin(_msgSender(), roleId);
+        }
+        _;
+    }
+
+    function initialize(address ozAccessManager, address registry) external initializer 
     {
-        // if size of the contract gets too large, this can be externalized which will reduce the contract size considerably
-        _accessManager = new AccessManager(address(this));
-        // this service required admin rights to access manager to be able to grant/revoke roles
-        _accessManager.grantRole(_accessManager.ADMIN_ROLE(), initialAdmin, 0);
+        require(ozAccessManager != address(0));
+        require(registry != address(0));
+
+        __AccessManaged_init(ozAccessManager);
 
-        __AccessManaged_init(address(_accessManager));
+        _accessManager = AccessManagerUpgradeableInitializeable(ozAccessManager);
+        _registry = IRegistry(registry);
+        _idNext = CUSTOM_ROLE_ID_MIN;
 
-        _createRole(RoleIdLib.toRoleId(_accessManager.ADMIN_ROLE()), ADMIN_ROLE_NAME, false, false);
-        _createRole(RoleIdLib.toRoleId(_accessManager.PUBLIC_ROLE()), PUBLIC_ROLE_NAME, false, false);
+        _createRole(ADMIN_ROLE(), ADMIN_ROLE_NAME, IAccess.Type.Core);
+        _createRole(PUBLIC_ROLE(), PUBLIC_ROLE_NAME, IAccess.Type.Core);
+
+        // assume `this` is already a member of ADMIN_ROLE
+        // assume msg.sender is instance service which is also member of ADMIN_ROLE
+        // assume instance service will renounce ADMIN_ROLE through ozAccessManager and should not be added to _roleMembers here
+        EnumerableSet.add(_roleMembers[ADMIN_ROLE()], address(this));
+        //EnumerableSet.add(_roleMembers[ADMIN_ROLE()], initialAdmin);
     }
 
     //--- Role ------------------------------------------------------//
-    function createGifRole(RoleId roleId, string memory name) external restricted() {
-        _createRole(roleId, name, false, true);
+    // ADMIN_ROLE
+    // assume all core roles are know at deployment time
+    // assume core roles are set and granted only during instance cloning
+    // assume core roles are never revoked or renounced -> core roles admin is never active after intialization
+    function createCoreRole(RoleId roleId, string memory name)
+        external
+        restricted()
+    {
+        _createRole(roleId, name, IAccess.Type.Core);
     }
+    // INSTANCE_SERVICE_ROLE
+    // assume gif roles can be revoked or renounced
+    // TODO who can be admin of gif role?
+    function createGifRole(RoleId roleId, string memory name, RoleId admin) 
+        external
+        restricted()
+    {
+        _createRole(roleId, name, IAccess.Type.Gif);
+        setRoleAdmin(roleId, admin);
+    }
+
+    // INSTANCE_OWNER_ROLE
```

Comment:
could be more then instance_owner_role

Reply:
may be, is it clear who else can create custom role?